app/src/rgb_underglow.c

```diff
@@ -313,63 +335,70 @@ int zmk_rgb_underglow_cycle_effect(int direction) {
 }
 
 int zmk_rgb_underglow_toggle() {
-    if (!led_strip)
-        return -ENODEV;
+    return state.on ? zmk_rgb_underglow_off() : zmk_rgb_underglow_on();
+}
+
+int zmk_rgb_underglow_set_hsb(struct zmk_led_hsb color) {
+    if (color.h > 360 || color.s > 100 || color.b > 100) {
+        return -ENOTSUP;
+    }
 
-    state.on = !state.on;
+    state.color = color;
 
-#if IS_ENABLED(CONFIG_ZMK_RGB_UNDERGLOW_EXT_POWER)
-    if (ext_power != NULL) {
-        int rc;
+    return 0;
+}
 
-        if (state.on) {
-            rc = ext_power_enable(ext_power);
-        } else {
-            rc = ext_power_disable(ext_power);
-        }
+struct zmk_led_hsb zmk_rgb_underglow_calc_hue(int direction) {
+    struct zmk_led_hsb color = state.color;
 
-        if (rc != 0) {
-            LOG_ERR("Unable to toggle EXT_POWER: %d", rc);
-        }
+    if (color.h == 0 && direction < 0) {
+        return color;
```

Comment:
I don't think this is sufficient, and it also doesn't support wrapping around. Instead of this, after updating the hue with `+= direction * CONFIG_ZMK_RGB_UNDERGLOW_HUE_STEP`, if the result is less than 0, add 360.